### ISSUE TITLE
Unable to create the cache directory

### ISSUE THREAD

**Issue by tribly** [firefly-iii/firefly-iii#2970]
**Bug description**
I am running Firefly III version 4.8.2, and my problem is:

Ever since upgrading to php7.4 I'm getting the following error:

```
2020/01/07 10:54:38 [error] 480162#480162: *33 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught RuntimeException: Unable to create the cache directory (/usr/share/webapps/firefly-iii/storage/framework/views/twig/58). in /usr/share/webapps/firefly-iii/vendor/twig/twig/src/Cache/FilesystemCache.php:57
Stack trace:
#0 /usr/share/webapps/firefly-iii/vendor/twig/twig/src/Environment.php(487): Twig\Cache\FilesystemCache->write('/usr/share/weba...', '<?php\n\nuse Twig...')
#1 /usr/share/webapps/firefly-iii/vendor/twig/twig/src/Environment.php(445): Twig\Environment->loadClass('__TwigTemplate_...', '/usr/share/weba...', NULL)
#2 /usr/share/webapps/firefly-iii/vendor/rcrowe/twigbridge/src/Bridge.php(75): Twig\Environment->loadTemplate('/usr/share/weba...', NULL)
#3 /usr/share/webapps/firefly-iii/vendor/rcrowe/twigbridge/src/Engine/Compiler.php(99): TwigBridge\Bridge->loadTemplate('/usr/share/weba...')
#4 /usr/share/webapps/firefly-iii/vendor/rcrowe/twigbridge/src/Engine/Twig.php(91): TwigBridge\Engine\Compiler->load('/usr/share/weba...')
#5 /usr/share/webapps/f...PHP message: PHP Fatal error:  Uncaught RuntimeException: Unable to create the cache directory (/usr/share/webapps/firefly-iii/storage/framework/views/twig/58). in /usr/share/webapps/firefly-iii/vendor/twig/twig/src/Cache/FilesystemCache.php:57
Stack trace:
#0 /usr/share/webapps/firefly-iii/vendor/twig/twig/src/Environment.php(487): Twig\Cache\FilesystemCache->write('/usr/share/weba...', '<?php\n\nuse Twig...')
#1 /usr/share/webapps/firefly-iii/vendor/twig/twig/src/Environment.php(445): Twig\Environment->loadClass('__TwigTemplate_...', '/usr/share/weba...', NULL)
#2 /usr/share/webapps/firefly-iii/vendor/rcrowe/twigbridge/src/Bridge.php(75): Twig\Environment->loadTemplate('/usr/share/weba...', NULL)
#3 /usr/share/webapps/firefly-iii/vendor/rcrowe/twigbridge/src/Engine/Compiler.php(99): TwigBridge\Bridge->loadTemplate('/usr/share/weba...')
#4 /usr/share/webapps/firefly-iii/vendor/rcrowe/twigbridge/src/Engine/Twig.php(91): TwigBridge\Engine\Compiler
```

The permissions are still correct and fpm is running as http:http. Unfortunately it's too much of a hassle to downgrade php to test if it's a 7.4 problem. Not using a docker image, installed via composer. Tried reinstalling again, didn't work.

**Steps to reproduce**
Upgrade to php7.4 maybe.

**Comment by JC5** [comment]
Clearly, the permissions are not correct. Make sure that the user that runs your app has write access to the cache directory and all the sub-directories.

**Comment by tribly** [comment]
Nginx is running as http:http, php-fpm also. Just reinstalled again, set the proper 775 permissions for `storage`, the owner/group to http. Doesn't work.

My other php applications are working without a problem. I'm pretty confused.

**Comment by JC5** [comment]
Sorry, but the error seems pretty unambiguous. I'm not sure what else it could be. Firefly III's code is not that special when it comes to directories.

**Comment by ctrl-b** [comment]
I tried to build the image locally but still getting the same error, It seems like a permissions error chown is not working either, probably wait for v5? I'll try to export/backup my data just in case

**Comment by JC5** [comment]
Oh that’s an edge case, nice find.

**Comment by tribly** [comment]
I moved the firefly dir to another location and now it's works for whatever strange reason.

**Comment by mlrcedjk** [comment]
Hi James and tribly,
in the last days I had the same issue mentioned by tribly. I configured my directory permissions properly and moved the storage directory to another location. For a few minutes this workaround solves the permissions problem. I set up firefly-iii with PHP 7.4.

As described in [Archwiki Nextcloud](https://wiki.archlinux.org/index.php/Nextcloud#Explicitly_permit_Nextcloud_directories_for_php-fpm) PHP 7.4 is hardened by default and I had to give permissions for firefly-iii paths explicitly in the systemd service.

1. `systemctl edit php-fpm.service`
2. Add the following lines (my firefly-iii instance is in /usr/share/webapps/)
`[Service]`
`ReadWritePaths = /usr/share/webapps/firefly-iii/storage`
3. `systemctl restart php-fpm.service`

Maybe this will help you and others.

Greetings,
mlrcedjk
